core/src/main/java/io/questdb/griffin/engine/functions/memoization/IntFunctionMemoizer.java

```diff
@@ -0,0 +1,81 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2024 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.memoization;
+
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.SymbolTableSource;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.functions.IntFunction;
+import io.questdb.griffin.engine.functions.UnaryFunction;
+
+public final class IntFunctionMemoizer extends IntFunction implements UnaryFunction {
```

Comment:
Memoization doesn't seem to play well with random access
```sql
select rnd_int()+1 x1, x1+1 x12 from long_sequence(3);
```
returns

| x1         | x12        |
| ---------- | ---------- |
| -629501265 | -629501264 |
| -96661474  | -96661473  |
| 1287899276 | 1287899277 |

while
```sql
select rnd_int()+1 x1, x1+1 x12 from long_sequence(3) order by x12;
```
returns all equal values

| x1          | x12         |
| ----------- | ----------- |
| -1947092543 | -1947092542 |
| -1947092543 | -1947092542 |
| -1947092543 | -1947092542 |

Reply:
fixed in https://github.com/questdb/questdb/pull/5778/commits/48ccbc2a36b6d87fd0c483746ee09de0eeee4d34

Reply:
ok, introduced random specific flag to materialize rnd functions in order by and other factories: [26779f4](https://github.com/questdb/questdb/pull/5778/commits/26779f4614c3a920102fe695d2c491901c70ad7e)

Reply:
That should work. Now we need to add some tests around this memoizer + random access.